flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java

```diff
@@ -0,0 +1,422 @@
+
+		return stateHandles;
+	}
+
+	/**
+	 * Removes a state handle from ZooKeeper.
+	 *
+	 * <p><stong>Important</stong>: this does not discard the state handle. If you want to
+	 * discard the state handle call {@link #discard(String)}.
+	 *
+	 * @param pathInZooKeeper Path of state handle to remove (expected to start with a '/')
+	 * @throws Exception
+	 */
+	public void remove(String pathInZooKeeper) throws Exception {
+		checkNotNull(pathInZooKeeper, "Path in ZooKeeper");
+
+		client
+				.delete()
+				.deletingChildrenIfNeeded()
+				.forPath(pathInZooKeeper);
+	}
+
+	/**
+	 * Removes a state handle from ZooKeeper asynchronously.
+	 *
+	 * <p><stong>Important</stong>: this does not discard the state handle. If you want to
+	 * discard the state handle call {@link #discard(String)}.
+	 *
+	 * @param pathInZooKeeper Path of state handle to remove (expected to start with a '/')
+	 * @param callback        The callback after the operation finishes
+	 * @throws Exception
+	 */
+	public void remove(String pathInZooKeeper, BackgroundCallback callback) throws Exception {
+		checkNotNull(pathInZooKeeper, "Path in ZooKeeper");
+		checkNotNull(callback, "Background callback");
+
+		client
+				.delete()
+				.deletingChildrenIfNeeded()
+				.inBackground(callback)
+				.forPath(pathInZooKeeper);
+	}
+
+	/**
+	 * Discards a state handle and removes it from ZooKeeper.
+	 *
+	 * <p><strong>Important</strong>: This method will only discard the state handle and not the
+	 * state itself. Don't forget to run custom cleanup code of the state, if necessary.
+	 *
+	 * <pre>
+	 * T state = get(path).getState();
+	 * state.discard(); // Custom clean up
+	 * discard(path)
+	 * </pre>
+	 *
+	 * <p>If you only want to remove the state handle in ZooKeeper call {@link #remove(String)}.
+	 *
+	 * @param pathInZooKeeper Path of state handle to discard (expected to start with a '/')
+	 * @throws Exception
+	 */
+	public void discard(String pathInZooKeeper) throws Exception {
+		checkNotNull(pathInZooKeeper, "Path in ZooKeeper");
+
+		StateHandle<T> stateHandle = get(pathInZooKeeper);
+
+		// Delete the state handle from ZooKeeper first
+		client
+				.delete()
+				.deletingChildrenIfNeeded()
+				.forPath(pathInZooKeeper);
+
+		// Discard the state handle only after it has been successfully deleted from ZooKeeper.
+		// Otherwise we might enter an illegal state after failures (with a state handle in
+		// ZooKeeper, which has already been discarded).
+		stateHandle.discardState();
+	}
+
+	/**
+	 * Discards all available state handles and removes them from ZooKeeper.
+	 *
+	 * <p><strong>Important</strong>: This method will only discard the state handles and not the
+	 * states itself. Don't forget to run custom cleanup code of the states, if necessary.
+	 *
+	 * <pre>
+	 * for (StateHandle<T> handle : getAll()) {
+	 *     handle.getState().discard(); // Custom clean up
+	 * }
+	 * discardAll()
+	 * </pre>
+	 *
+	 * @throws Exception
```

Comment:
Resolved. The root problem here is that both the state handle and curator APIs declare Exception on most operations w/o more details.